Added using-blob-storage-with-abp article (#19733)
abpframework/abp: docs/en/Community-Articles/2024-04-19-using-blob-storage-with-abp/post.md

# Using Blob Storage with ABP

>  ABP Framework is a comprehensive development framework for modern web applications and microservice architectures. This framework offers its users the possibility to quickly build scalable and modular applications. An advantage of this framework is the flexibility to integrate different cloud services or infrastructure components. 

Reviewer: In this paragraph, you are only describing the ABP Framework but do not link the paragraph with the article. Therefore, I think you should first plainly describe the framework and then you should mention that ABP Framework provides Blob Storing infrastructure to easily integrate into your application and provides multiple pre-built providers.

So, I think you may update this paragraph as below:

```suggestion
 ABP Framework is a comprehensive development framework for modern web applications and microservice architectures. ABP Framework [provides an abstraction to work with BLOBs and provides some pre-built storage providers](https://docs.abp.io/en/abp/latest/Blob-Storing) that you can easily integrate into your application. In this article, I will show you how to store BLOBs in a relational or non-relational database by using the [Database Provider](https://docs.abp.io/en/abp/latest/Blob-Storing-Database).
```

You can update the paragraph with your sentences, but keep that in mind, adding links for further information is a good practice, so the readers can check the official document if they want to learn more.

## What is Blob Storage ?

   Blob Storage is a cloud storage service for storing large amounts of data. This service is ideal for storing files, images or other types of data at high scale. The advantages of Blob Storage include scalability, high performance, durability and low cost. In addition, Blob Storage's APIs make it easy to access and manage data. The Blob Storage integration with ABP Framework offers a powerful combination to effectively meet the big data storage needs of your applications. This integration can help you optimize storage costs while improving the performance of your application.

   ![Blob Stroge](./images/blob-storage.png)

## How to use Blob Storage ?

   Nowadays, storing large files such as users' profile pictures has become an important requirement in web applications. Storing such data in the database can negatively impact performance and increase the database size. To solve this problem, Blob storage services can be used. ABP Framework provides a powerful solution to handle such scenarios. So, how do we store user profile pictures with Blob Storage using ABP Framework?

> - #### Step 1: Blob Container Identification

Reviewer: ```suggestion
- #### Step 1: Configure the Blob Container
```

This might be a better title for this section.

The first step is to define a Blob Container for storing profile pictures. The Blob Container represents the storage space that will be used to store the profile pictures.

>    ![Step 1](./images/step1.png)

Reviewer: If you are showing a code-block to the readers, they should be able to directly copy-paste the code-blocks and try it instantly within their applications. Therefore, showing the code in the code-blocks in markdown is a good practise. Thus, please update this and all images with the relevant code-blocks. (Also, you can use : before the previous sentence of the code-block):

````suggestion
The first step is to define a Blob Container for storing profile pictures. The Blob Container represents the storage space that will be used to store the profile pictures:

```csharp
public class ProfilePictureContainer : IBlobContainer
{
    public string Name => "profile-pictures";
}
```

````

Please also update other images that contain code-blocks.

---

Also, you may mention that the `BlobContainerNameAttribute` can be used to set the container name (https://docs.abp.io/en/abp/latest/Blob-Storing#typed-iblobcontainer)

> - #### Step 2: Create ProfileAppService

Reviewer: ```suggestion
- #### Step 2: Create the ProfileAppService (Saving & Reading BLOBs)
```

This might be a better title for this section, but it can stay as it is, it's totally up to you.

> Create the ProfileAppService class and derive it from the ApplicationService class. This class will perform the necessary operations to store and retrieve profile pictures.

Reviewer: For the class, property, and method names, you can use the (`) character:

```suggestion
Create the `ProfileAppService` class and derive it from the `ApplicationService` class. This class will perform the necessary operations to store and retrieve profile pictures.
```

Please wrap the special names (code-related) with ` character. Like in the example above.

   ![Step 2](./images/step2.png)

> - #### Step 3: Blob Container Connection

Reviewer: ```suggestion
- #### Step 3: Inject the `IBlobContainer` Service
```

This might be a better title for this section, "Blob Container Connection" can be misleading as a title for the section.

> In the constructor method, inject the dependency into the IBlobContainer interface. This interface is used to interact with the Blob Container.

Reviewer: > "In the constructor method, inject the dependency into the IBlobContainer interface." 

It occurred to me that this sentence is not right (not in the correct form) and I think you can update it as below:

```suggestion
Inject the `IBlobContainer` service, in the constructor of the `ProfileAppService` class. The `IBlobContainer` is the main interface to store and read BLOB and is used to interact with the container.
```

   ![Step 3](./images/step3.png)

- #### Step 4: Save Profile Picture

The SaveProfilePictureAsync method is used to store the user's profile picture. A unique name is generated based on the user's credentials and the profile picture byte array with this name is saved in the Blob Container.

   ![Step 4](./images/step4.png)

- #### Step 5: Getting Profile Picture

The GetProfilePictureAsync method is used to get the user's profile picture. A profile picture byte array is retrieved from the Blob Container with a specified name based on the user's credential.

   ![Step 5](./images/step5.png)

Finally, add controls in the user interface that will allow users to upload and view their profile pictures. These controls will perform the operations by calling the corresponding methods in the ProfileAppService class.

These steps cover the basic steps to store user profile pictures with Blob Storage using the ABP Framework. [Check out the documentation for more information.](https://docs.abp.io/en/abp/latest/Blob-Storing)

> ## What are the advantages/disadvantages of keeping in Blob Storage database? 

Reviewer: ```suggestion
## What are the Advantages/Disadvantages of Keeping the BLOBs in a Database? 
```

- A Single Storage Location: Storing blob data in the database allows you to collect all data in a single storage location. This simplifies the management of data and increases data integrity. 

- Advanced Security Controls: Database systems often offer advanced security controls. Storing blob data in a database allows you to take advantage of these security features and ensures that data is accessed by authorized users. 

#### Disadvantages: 

- Performance Issues: Storing blob data in a database can negatively impact database performance. Oversized blob data can slow down query processing and reduce database performance. 

- Storage Space Issue: Storing blob data in the database can increase the size of the database and require more storage space. This can increase storage costs and complicate infrastructure requirements. 

- Backup and Recovery Challenges: Storing blob data in a database can make backup and recovery difficult. The large size of blob data can make backup and recovery time-consuming and data recovery difficult. 

> ## Information about Blob Storage entagrations  

Reviewer: It's a [good practise to having a paragraph](https://developers.google.com/tech-writing/two/large-docs) after the title to pre-explain the section:

```suggestion
## Other Blob Storage Providers

ABP Framework provides some other [pre-built storage providers](https://docs.abp.io/en/abp/latest/Blob-Storing#blob-storage-providers) besides the [Database Provider](https://docs.abp.io/en/abp/latest/Blob-Storing-Database) shown in this article:
```
